arwen/contexts/blockchain.go

```diff
@@ -47,7 +47,15 @@ func (context *blockchainContext) AccountExists(address []byte) bool {
 func (context *blockchainContext) GetBalance(address []byte) []byte {
 	outputAccount, isNew := context.host.Output().GetOutputAccount(address)
 	if !isNew {
-		return outputAccount.Balance.Bytes()
+		if outputAccount.Balance == nil {
+			balance, err := context.blockChainHook.GetBalance(address)
+			if err != nil {
+				return big.NewInt(0).Bytes()
```

Comment:
Must leave it to `nil`, in order to know that the balance has not been actually retrieved from the `BlockchainHook`. Otherwise, the following may happen:

1. SmartContract does `transferValue()` and sends 5 ERD to Alice
2. The "Alice" OutputAccount is created in the cache of the `outputContext`, with `BalanceDelta = 5`.
3. SmartContract does `getBalance(Alice)`, which directly returns 5 ERD from cache, regardless of the real amount in Alice's account.

Setting `Balance` to `nil` indicates that the cached `OutputAccount` does not hold the real balance, which must be retrieved on demand.